Убрать env файлы из репозитория (#55)

birthdaysgift opened the pull request:
На данный момент наш репозиторий отслеживает следующие .env файлы:
- `envs/local/dev/.env`
- `envs/local/test/.env`
- `envs/ci/lint/.env`
- `envs/ci/test/.env`

Для локальной разработки это может быть не очень удобно, т.к. если кто-то захочет изменить что-нибудь в конфигурации, то эти изменения будут видны в диффе. Например, у кого-то может быть занят порт для постгреса (5432) и он локально захочет поменять его на другой.

Такие же проблемы возникнут и для прод окружения, т.к. если мы не будем коммитить .env файл для прода, то мы, во-первых, нарушим консистентность со всеми остальными окружениями, а во-вторых, у нас не будет удобного примера .env файла и нам придётся его копировать из какого-либо окружения. А разные окружения могут содержать разный набор параметров. Также на проде могут быть параметры специфичные только для прода, и надо будет помнить о том, чтобы их добавить. А если его всё таки закоммитить .env файл, то поскольку его всё равно придётся менять на проде (добавлять секреты), то он окажется в диффе на прод машине (мы планируем использовать гит на хост машине, чтобы забирать последнюю версию репы для деплоя). Это затруднит работу с репозиторием на прод машине - например, наличие диффа усложнит ребейзы ветки на последнюю версию. - Следовательно, эта проблема блокирует задачу #51.

Также при этом не хотелось бы заставлять разработчика вручную прописывать переменные для каждого окружения, т.к. со временем их количество будет возрастать. Можно было бы сделать что-то вроде `.env.example`, в котором будут перечисленны дефолтные значения для каждого окружения, но в таком случае разработчику всё равно придётся для каждого окружения руками создавать `.env` файл и копировать туда значения.

Таким образом у этого вопроса есть три аспекта:
- мы хотим иметь возможность кастомизировать `.env` конфиги, чтобы они не валялись в диффе
- мы хотим иметь дефолтных рабочих значений для каждого окружения (секреты на проде будут исключением)
- мы хотим чтобы дефолтные значения можно было удобно устанавливать в каждое окружения - свести к минимуму ручную работу

Кажется, эти три проблемы можно решить, если для каждого окружения написать небольшой баш скрипт, который бы создавал `.env` файл в папке нужного окружения и записывал туда дефолтные значения. При этом сами `.env` файлы можно удалить из репозитория и заигнорить их в `.gitignore`. Таким образом наши дефолтные значения будут всегда прописаны в баш скрипте и закоммичены в репозиторий, но при этом генерируемый .env файл будет исключен из репозитория, что позволит его кастомизировать. А также настройка дефолтных переменных для каждого окружения будет заключаться всего лишь в одной команде запуска баш скрипта.

Единственная проблема будет в том, что в рамках CI нам нет необходимости кастомизировать переменные окружения, но для сохранения консистентности всё равно надо будет сделать такой же скрипт и запускать его перед запуском докер контейнеров. Но мне не кажется, что это НЕ будет каким-либо значимым оверхедом, поэтому можно закрыть на это глаза в пользу консистентности.

В рамках этой задачи необходимо создать такие скрипты для каждого из существующих окружений, добавить команду запуска этих скриптов в `README.md`, добавить использование этой команды в воркфлоу для CI, удалить `.env` файлы из репозитория и заигнорить их в `.gitignore`.

Rey-Lane reviewed (approved):
Можно добавить пару строк об этом в ридми, а то информация о том, что это за скрипты и для чего они нужны останется в недрах гита, не больше. Ну и на CI что-то кешем pytest)

birthdaysgift commented:
> Можно добавить пару строк об этом в ридми, а то информация о том, что это за скрипты и для чего они нужны

Да, конечно, это же есть в описании тикета.

> Ну и на CI что-то кешем pytest

Это причина почему пр сейчас в драфте, работаю над этим.

birthdaysgift committed:
Убрать env файлы из репозитория (#55)

На данный момент наш репозиторий отслеживает следующие .env файлы:
- `envs/local/dev/.env`
- `envs/local/test/.env`
- `envs/ci/lint/.env`
- `envs/ci/test/.env`

Для локальной разработки это может быть не очень удобно, т.к. если кто-то захочет изменить что-нибудь в конфигурации, то эти изменения будут видны в диффе. Например, у кого-то может быть занят порт для постгреса (5432) и он локально захочет поменять его на другой.

Такие же проблемы возникнут и для прод окружения, т.к. если мы не будем коммитить .env файл для прода, то мы, во-первых, нарушим консистентность со всеми остальными окружениями, а во-вторых, у нас не будет удобного примера .env файла и нам придётся его копировать из какого-либо окружения. А разные окружения могут содержать разный набор параметров. Также на проде могут быть параметры специфичные только для прода, и надо будет помнить о том, чтобы их добавить. А если его всё таки закоммитить .env файл, то поскольку его всё равно придётся менять на проде (добавлять секреты), то он окажется в диффе на прод машине (мы планируем использовать гит на хост машине, чтобы забирать последнюю версию репы для деплоя). Это затруднит работу с репозиторием на прод машине - например, наличие диффа усложнит ребейзы ветки на последнюю версию. - Следовательно, эта проблема блокирует задачу #51.

Также при этом не хотелось бы заставлять разработчика вручную прописывать переменные для каждого окружения, т.к. со временем их количество будет возрастать. Можно было бы сделать что-то вроде `.env.example`, в котором будут перечисленны дефолтные значения для каждого окружения, но в таком случае разработчику всё равно придётся для каждого окружения руками создавать `.env` файл и копировать туда значения.

Таким образом у этого вопроса есть три аспекта:
- мы хотим иметь возможность кастомизировать `.env` конфиги, чтобы они не валялись в диффе
- мы хотим иметь дефолтных рабочих значений для каждого окружения (секреты на проде будут исключением)
- мы хотим чтобы дефолтные значения можно было удобно устанавливать в каждое окружения - свести к минимуму ручную работу

Кажется, эти три проблемы можно решить, если для каждого окружения написать небольшой баш скрипт, который бы создавал `.env` файл в папке нужного окружения и записывал туда дефолтные значения. При этом сами `.env` файлы можно удалить из репозитория и заигнорить их в `.gitignore`. Таким образом наши дефолтные значения будут всегда прописаны в баш скрипте и закоммичены в репозиторий, но при этом генерируемый .env файл будет исключен из репозитория, что позволит его кастомизировать. А также настройка дефолтных переменных для каждого окружения будет заключаться всего лишь в одной команде запуска баш скрипта.

Единственная проблема будет в том, что в рамках CI нам нет необходимости кастомизировать переменные окружения, но для сохранения консистентности всё равно надо будет сделать такой же скрипт и запускать его перед запуском докер контейнеров. Но мне не кажется, что это НЕ будет каким-либо значимым оверхедом, поэтому можно закрыть на это глаза в пользу консистентности.

В рамках этой задачи необходимо создать такие скрипты для каждого из существующих окружений, добавить команду запуска этих скриптов в `README.md`, добавить использование этой команды в воркфлоу для CI, удалить `.env` файлы из репозитория и заигнорить их в `.gitignore`.
